src/View/Helper/PaginatorHelper.php

```diff
@@ -502,88 +502,95 @@ public function sort(string $key, $title = null, array $options = []): string
-            'direction' => $paging['direction'],
-        ];
-
-        if (!empty($this->_config['options']['url'])) {
-            $key = implode('.', array_filter(['options.url', Hash::get($paging, 'scope', null)]));
-            $url = array_merge($url, Hash::get($this->_config, $key, []));
-        }
-
-        $url = array_filter($url, function ($value) {
-            return $value || is_numeric($value) || $value === false;
-        });
-        $url = array_merge($url, $options);
+        $options += array_intersect_key(
+            $paging,
+            ['page' => null, 'limit' => null, 'sort' => null, 'direction' => null]
+        );
 
-        if (!empty($url['page']) && $url['page'] === 1) {
-            $url['page'] = false;
+        if (!empty($options['page']) && $options['page'] === 1) {
+            $options['page'] = null;
         }
 
-        if (isset($paging['sortDefault'], $paging['directionDefault'], $url['sort'], $url['direction']) &&
-            $url['sort'] === $paging['sortDefault'] &&
-            strtolower($url['direction']) === strtolower($paging['directionDefault'])
+        if (isset($paging['sortDefault'], $paging['directionDefault'], $options['sort'], $options['direction']) &&
+            $options['sort'] === $paging['sortDefault'] &&
+            strtolower($options['direction']) === strtolower($paging['directionDefault'])
         ) {
-            $url['sort'] = $url['direction'] = null;
+            $options['sort'] = $options['direction'] = null;
         }
 
         if (!empty($paging['scope'])) {
             $scope = $paging['scope'];
             $currentParams = $this->_config['options']['url'];
 
-            if (isset($url['#'])) {
-                $currentParams['#'] = $url['#'];
-                unset($url['#']);
-            }
-
-            // Merge existing query parameters in the scope.
             if (isset($currentParams['?'][$scope]) && is_array($currentParams['?'][$scope])) {
-                $url += $currentParams['?'][$scope];
-                unset($currentParams['?'][$scope]);
-            }
-            $url = [$scope => $url] + $currentParams;
-            if (empty($url[$scope]['page'])) {
-                unset($url[$scope]['page']);
+                $options += $currentParams['?'][$scope];
             }
+            $options = [$scope => $options];
+        }
+
+        if (!empty($this->_config['options']['url'])) {
+            $key = implode('.', array_filter(['options.url', Hash::get($paging, 'scope', null)]));
+            $url = Hash::merge($url, Hash::get($this->_config, $key, []));
+        }
+
+        if (!isset($url['?'])) {
+            $url['?'] = [];
+        }
+
+        if (!empty($this->_config['options']['routePlaceholders'])) {
+            $placeholders = array_flip($this->_config['options']['routePlaceholders']);
+            $url += array_intersect_key($options, $placeholders);
+            $url['?'] += array_diff_key($options, $placeholders);
```

Comment:
Should the placeholder parameters be added to the query string as well?

Reply:
I don't think so. Those using the paging params as URL fragment do so because they want prettier URLs without query string.